### PR TITLE
os/bluestore: partial reshard support

### PR DESCRIPTION
**Pull request by liewegas**
Reshard only a subset of the extent map.  If a single extent gets big, we will reshard only
it.  If one gets small, we will combine with the smaller of our two neighbors.

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -1906,7 +1898,14 @@ void BlueStore::ExtentMap::reshard(
   unsigned max_blob_end = 0;
   for (auto& e: extent_map) {
     dout(30) << " extent " << e << dendl;
-    assert(!e.blob->is_spanning());
+
+    // unspan spanning blobs
+    if (e.blob->is_spanning()) {
+      spanning_blob_map.erase(e.blob->id);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Can we simply call spanning_blob_map.clear() after the loop?

Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: not anymore.. we only want to unspan the blobs within the given range

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -1985,43 +2065,50 @@ void BlueStore::ExtentMap::reshard(
...
+	    dout(20) << __func__ << "    adding spanning " << *b << dendl;
 	  }
-	} else {
-	  must_span = true;
 	}
-	if (must_span) {
-	  b->id = bid++;
-	  spanning_blob_map[b->id] = b;
-	  dout(20) << __func__ << "    adding spanning " << *b << dendl;
+      } else {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: else -> else if(e->blob->is_spanning())

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -1951,17 +1978,61 @@ void BlueStore::ExtentMap::reshard(
...
+    shards.erase(shards.begin() + si_begin, shards.begin() + si_end);
+    sv.insert(
+      sv.begin() + si_begin,
+      new_shard_info.begin(),
+      new_shard_info.end());
+    shards.insert(shards.begin() + si_begin, new_shard_info.size(), Shard());
+    unsigned n = sv.size();
+    si_end = si_begin + new_shard_info.size();
+    for (unsigned i = si_begin; i < si_end; ++i) {
+      shards[i].offset = sv[i].offset;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Wondering if we need offset field in Shard at all, isn't one in shard_info enough?

File: src/os/bluestore/BlueStore.h [hunk 4]
```diff
@@ -633,14 +662,31 @@ class BlueStore : public ObjectStore,
     struct Shard {
       bluestore_onode_t::shard_info *shard_info;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: Suggest to set to nullptr

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -1951,17 +1978,61 @@ void BlueStore::ExtentMap::reshard(
       max_blob_end = be;
     }
   }
-  onode->onode.extent_map_shards.swap(new_shard_info);
+  if (new_shard_info.empty() && (si_begin > 0 ||
+				 si_end < shards.size())) {
+    // we resharded a partial range; we must produce at least one output
+    // shard
+    new_shard_info.emplace_back(bluestore_onode_t::shard_info());
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: You might want to implement shard_info(offset) ctor and apply it here and throughout?

**Comment by liewegas** [comment]
Updated, with several bugs fixed.  And rebased now that use_tracker is merged.  Please take a look!

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -1823,21 +1822,22 @@ void BlueStore::ExtentMap::update(KeyValueDB::Transaction t,
...
 	  // avoid resharding the trailing shard, even if it is small
-	  if (n != shards.end() &&
-	      len < g_conf->bluestore_extent_map_shard_min_size) {
+	  else if (n != shards.end() &&
+		   len < g_conf->bluestore_extent_map_shard_min_size) {
 	    // we are small; combine with a neighbor
 	    if (p == shards.begin() && endoff == OBJECT_MAX_SIZE) {
 	      // we are an only shard
 	      request_reshard(0, OBJECT_MAX_SIZE);
+	      return;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: We'll exist the loop at this iteration and return the function anyway hence just a noice

File: src/os/bluestore/BlueStore.cc [hunk 7]
```diff
@@ -1823,21 +1822,22 @@ void BlueStore::ExtentMap::update(KeyValueDB::Transaction t,
...
 	    } else if (p == shards.begin()) {
 	      // combine with next shard
 	      request_reshard(p->offset, endoff + 1);
 	    } else if (endoff == OBJECT_MAX_SIZE) {
 	      // combine with previous shard
 	      request_reshard(prev_p->offset, endoff);
+	      return;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: same here

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-02-07_18:56:54-rados-master---basic-smithi/

**Pull request merged**